Add a flag to overwrite existing output directory

lgp171188 opened the pull request:
Fix issue #493 and allow the user to alter the default behaviour
behaviour of not overwriting the contents of the output directory
if it already exists. Fix issue #475 by removing the conflicting
use of 'fail_if_exists' arguments and refactoring it.

lgp171188 commented:
cc @hackebrot @schacki @ionelmc

hackebrot commented on cookiecutter/main.py:
```diff
@@ -52,7 +52,8 @@ def expand_abbreviations(template, config_dict):
 
 
-def cookiecutter(template, checkout=None, no_input=False, extra_context=None):
+def cookiecutter(template, checkout=None, no_input=False,
```
I am afraid this breaks backwards compatibility :worried: 

We should append the new kwarg at the very end.

lgp171188 replied:
The reason I inserted it in the middle rather than at the end is that in ``cookiecutter/cli.py``, which is the only place this function is called, this function is invoked in this way - ``cookiecutter(template, checkout, no_input)``. This implies ``extra_context`` is an optional argument. But the flag ``overwrite_if_exists`` is always passed to all the functions that require it and putting it at the end will require code change like one of these 2 ways, which don't look that good imho.
    
    # first way
    extra_context = None
    cookiecutter(template, checkout, no_input, extra_context, overwrite_if_exists)

    # second way
    cookiecutter(template, checkout, no_input, overwrite_if_exists=overwrite_if_exists)

Please suggest better ways to deal with this, if any.

ionelmc replied:
There might still be external code depending on the position of arguments. Quite unfortunate as Python 2.x doesn't have keyword-only arguments.

lgp171188 replied:
@ionelmc @hackebrot Okay, I will wait for all the other review comments, if any, and fix this along with the other issues pointed out.

hackebrot replied:
I would most certainly go for kwargs as the default value for ``extra_context`` *could* change in the future (I have doubts it wil :wink:).

ionelmc commented:
Anyone still working on this?

hackebrot commented:
Nope, I am not. I'm still in favor of using kwargs :smirk:

pydanny commented:
:-1:  on breaking backwards compatibility. 

FWIW cookiecutter is used in a lot of places and packages, including as an imported package. If we accept this PR as it stands the issues board will be filled with help requests.

lgp171188 commented:
@pydanny I will fix the backward compatibility issues pointed out by @hackebrot. I was just waiting for all the review comments which would require a change so that I could do all of them in one shot. Will make the suggested changes and get back.

lgp171188 commented:
@hackebrot I have updated the pull request not to break the backward compatibility and also rebased on top of the latest changes in the master branch. Please review. /cc @ionelmc 

I also see a new replay feature added to the codebase. Not sure how ``--overwrite-if-exists`` should work with it. At the moment, I have made cookiecutter throw an error if it ``-f`` or ``--overwrite-if-exists`` is used along with replay since I am not sure about it.

What about defaulting to ``--overwrite-if-exists`` when using replay?

lgp171188 commented:
Ping @hackebrot :)

hackebrot commented:
Hi @lgp171188! Apologies for my late reply. :no_mouth: 

I think it should be possible to **replay** and **overwrite** at the same time.

hackebrot commented on cookiecutter/main.py:
```diff
@@ -70,8 +70,8 @@ def expand_abbreviations(template, config_dict):
-        template,
-        checkout=None, no_input=False, extra_context=None, replay=False):
+        template, checkout=None, no_input=False, extra_context=None,
+        replay=False, **kwargs):
```
I'd prefer to use ``overwrite_if_exists=False`` instead of ``**kwargs``.

pydanny replied:
I agree. Explicit over implicit!

lgp171188 replied:
@pydanny @hackebrot It looks like misinterpreted the previous comments on using kwargs and thought that I should use the ``**kwargs`` structure. My bad. Will fix that :)

pydanny commented:
@lgp171188 This looks great. If you could just change the use of `kwargs` to an explicit argument, then I think this is ready to merge in.

Thank you for all your hard work!

lgp171188 commented:
Details of the most recent changes to this PR:
* Make ``overwrite_if_exists`` as an explicit argument instead of using it via the implicit ``**kwargs`` argument
* Make the ``overwrite-if-exists`` flag independent of the replay feature as suggested by @hackebrot

hackebrot commented on tests/test_cli.py:
```diff
@@ -101,11 +102,56 @@ def test_cli_exit_on_noinput_and_replay(mocker):
     )
+
+@pytest.mark.usefixtures('remove_fake_project_dir')
+def test_cli_exit_on_overwrite_if_exists_and_replay(mocker):
```
This test is not valid anymore as we do support using both options simultaneously.

hackebrot replied:
Or we should at least rename it :smile_cat:

lgp171188 replied:
@hackebrot I left this test intact because I felt that it was good to test both the options together and verify that the command exits normally.

hackebrot replied:
I see. :no_mouth: 

The name is a bit misleading, I'm afraid. I thought it is supposed to test that **replay** and **overwrite** cause a the cli to exit without running **cookiecutter** (see ``test_cli_exit_on_noinput_and_replay`` which asserts for an error exit code).

lgp171188 replied:
Then is it okay to rename the test to ``test_cli_successful_exit_on_overwrite_if_exists_and_replay`` or maybe ``test_cli_does_not_exit_with_failure_on_overwrite_if_exists_and_replay``? :-)

hackebrot commented on tests/test_cli.py:
```diff
@@ -101,11 +102,56 @@ def test_cli_exit_on_noinput_and_replay(mocker):
+
+@pytest.mark.usefixtures('remove_fake_project_dir')
+def test_cli_overwrite_if_exists_when_output_dir_does_not_exist():
+    if os.path.isdir('fake-project'):
```
I think we should move this code into a fixture.

lgp171188 replied:
I did not move it into a fixture because there is no other test as of now which uses similar code. Will move it to a fixture if you strongly feel so.

There is already a ``fin_remove_fake_project_dir`` fixture function but it is a nested function and set as a finalizer. So there is no scope to reuse it.

hackebrot commented:
Many thanks for your work @lgp171188! :bow:

pydanny commented:
Great work @lgp171188! :+1:
